File: scala/Solver/Solver.scala
package Solver
import com.microsoft.z3.*
import Undeterminable.tree
import Parsing.{Expr, Stmt}
import Lexing.Token
import scala.util.boundary

/**
 * Z3ソルバーとの統合
 * 
 * 制約解決により、トークナイズ・パース候補を絞り込む
 */
object Solver:
  
  /**
   * Z3コンテキストの作成
   * リソース管理のため、using パターンで使用すること
   */
  def createContext(): Context =
    val cfg = new java.util.HashMap[String, String]()
    cfg.put("model", "true")
    new Context(cfg)
  
  /**
   * デリミタのペア定義
   */
  val delimiterPairs: Map[String, String] = Map(
    "(" -> ")",
    "{" -> "}",
    "[" -> "]"
  )
  
  /**
   * トークン列のデリミタバランスを検証
   * 
   * Z3を使って以下を検証：
   * 1. 開き括弧と閉じ括弧の数が一致
   * 2. ネストが正しい（閉じが先に来ない）
   * 
   * @param tokens トークン列
   * @return バランスが取れているか
   */
  def checkDelimiterBalance(tokens: List[Token]): Boolean =
    scala.util.Using.resource(createContext()) { ctx =>
      val solver = ctx.mkSolver()
      
      // 各デリミタペアについて検証
      delimiterPairs.foreach { case (open, close) =>
        val openCount = tokens.count {
          case Token.Delim(d) if d == open => true
          case _ => false
        }
        val closeCount = tokens.count {
          case Token.Delim(d) if d == close => true
          case _ => false
        }
        
        // Z3制約：開きと閉じの数が等しい
        val openVar = ctx.mkIntConst(s"open_$open")
        val closeVar = ctx.mkIntConst(s"close_$close")
        
        solver.add(ctx.mkEq(openVar, ctx.mkInt(openCount)))
        solver.add(ctx.mkEq(closeVar, ctx.mkInt(closeCount)))
        solver.add(ctx.mkEq(openVar, closeVar))
      }
      
      // ネストの正しさも検証
      if !checkNestingOrder(tokens) then
        return false
      
      // Z3で検証
      solver.check() == Status.SATISFIABLE
    }
  
  /**
   * デリミタのネスト順序を検証（Z3不要の単純チェック）
   * 
   * 閉じ括弧が対応する開き括弧より前に来ないことを確認
   */
  private def checkNestingOrder(tokens: List[Token]): Boolean =
    val stack = scala.collection.mutable.Stack[String]()
    
    tokens.foreach {
      case Token.Delim(d) =>
        if delimiterPairs.contains(d) then
          // 開き括弧
          stack.push(d)
        else if delimiterPairs.values.toSet.contains(d) then
          // 閉じ括弧
          if stack.isEmpty then
            return false
          val open = stack.pop()
          if delimiterPairs.get(open) != Some(d) then
            return false
      case _ => // 他のトークンは無視
    }
    
    // 全て閉じられている必要がある
    stack.isEmpty
  
  /**
   * 識別子の命名規則を検証
   * 
   * @param tokens トークン列
   * @return 命名規則を満たすか
   */
  def checkIdentifierRules(tokens: List[Token]): Boolean =
    tokens.forall {
      case Token.Ident(name) =>
        // 識別子は英字またはアンダースコアで始まる
        name.headOption.exists(c => c.isLetter || c == '_')
      
      case Token.Number(num) =>
        // 数値は数字のみ
        num.forall(_.isDigit)
      
      case _ => true
    }
  
  /**
   * トークン列が全ての制約を満たすか検証
   * 
   * @param tokens トークン列
   * @return 制約を満たすか
   */
  def checkTokenConstraints(tokens: List[Token]): Boolean =
    checkDelimiterBalance(tokens) && checkIdentifierRules(tokens)
  
  /**
   * トークンツリーをZ3制約で枝刈り
   * 
   * 各パスに対して制約を検証し、違反するパスを削除
   * 
   * @param tokenTree トークンツリー
   * @return 枝刈り後のツリー
   */
  def pruneTokenTree(tokenTree: tree[Token]): tree[Token] =
    val paths = tokenTree.flattenPaths.map(_.toList)
    
    // 各パスを検証
    val validPaths = paths.filter(checkTokenConstraints)
    
    if validPaths.isEmpty then
      tree.DeadEnd
    else
      // 有効なパスから新しいツリーを構築
      rebuildTree(validPaths.toList)
  
  /**
   * トークン列のリストからツリーを再構築
   * 
   * TODO: より効率的な実装
   * 現在は単純に全パスを列挙する実装
   */
  private def rebuildTree(paths: List[List[Token]]): tree[Token] =
    if paths.isEmpty then
      tree.DeadEnd
    else
      tree.fork(paths.to(List).map(buildPath))
  
  private def buildPath(tokens: List[Token]): tree[Token] =
    tokens match
      case Nil => tree.Node(Vector.empty, List.empty)
      case head :: tail => tree.single(head, buildPath(tail))
  
  /**
   * AST制約の検証
   * 
   * 例：スコープ解決、名前の一意性など
   * 
   * @param stmts ステートメント列
   * @return 制約を満たすか
   */
  def checkASTConstraints(stmts: List[Stmt]): Boolean =
    // TODO: 実装
    // 1. 変数のスコープ検証
    // 2. 関数定義の一意性
    // 3. 制約式の検証
    
    ??? // 暫定:NotImplementedError
  
  /**
   * 制約式をSMT-LIB形式に変換
   * 
   * romanescoの制約式 → Z3で検証可能な形式
   * 
   * @param expr 制約式
   * @param ctx Z3コンテキスト
   * @return Z3の論理式
   */
  def exprToZ3(expr: Expr, ctx: Context): BoolExpr =
    // TODO: 実装
    // 1. 二項演算の変換
    // 2. 関数呼び出しの変換
    // 3. 制約の合成
    
    ??? // 暫定:NotImplementedError

/**
 * 制約の種類
 */
enum ConstraintType:
  case Delimiter    // デリミタのバランス
  case Identifier   // 識別子命名規則
  case Scope        // スコープ解決
  case Type         // 型制約（将来）
  case Custom       // ユーザー定義制約

/**
 * 制約違反のエラー情報
 */
case class ConstraintViolation(
  constraintType: ConstraintType,
  message: String,
  location: Option[Int] = None
)

File: Python/romanesco/solver.py
import z3
from typing import List, Set
from romanesco import undeterminable
from romanesco.lexing import Token, Delim, Ident, Number
from romanesco.parsing import Expr, Num, Var, Call, Block

# ======================================
# Solver (Z3 / Constraints)
# ======================================

delimiter_pairs = {
    "(": ")",
    "{": "}",
    "[": "]"
}

def check_delimiter_balance(tokens: List[Token]) -> bool:
    if not check_nesting_order(tokens):
        return False

    ctx = z3.Context()
    solver = z3.Solver(ctx=ctx)
    
    for open_char, close_char in delimiter_pairs.items():
        open_count = sum(1 for t in tokens if isinstance(t, Delim) and t.lexeme == open_char)
        close_count = sum(1 for t in tokens if isinstance(t, Delim) and t.lexeme == close_char)
        
        open_var = z3.Int(f"open_{open_char}", ctx=ctx)
        close_var = z3.Int(f"close_{close_char}", ctx=ctx)
        
        solver.add(open_var == open_count)
        solver.add(close_var == close_count)
        solver.add(open_var == close_var)
        
    return solver.check() == z3.sat

def check_nesting_order(tokens: List[Token]) -> bool:
    stack: List[str] = []
    close_chars = set(delimiter_pairs.values())
    
    for t in tokens:
        if isinstance(t, Delim):
            d = t.lexeme
            if d in delimiter_pairs:
                stack.append(d)
            elif d in close_chars:
                if not stack: return False
                open_char = stack.pop()
                if delimiter_pairs.get(open_char) != d: return False
    return len(stack) == 0

def check_identifier_rules(tokens: List[Token]) -> bool:
    for t in tokens:
        if isinstance(t, Ident):
            if not t.lexeme: return False
            first = t.lexeme[0]
            if not (first.isalpha() or first == '_'): return False
        elif isinstance(t, Number):
            if not t.lexeme.isdigit(): return False
    return True

def check_token_constraints(tokens: List[Token]) -> bool:
    return check_delimiter_balance(tokens) and check_identifier_rules(tokens)

def check_ast_constraints(exprs: List[Expr]) -> bool:
    # 1. Simple Scope Validation
    defined_vars = {"true", "false", "and", "or", "seq", "=", "+", "-", "*", "/", ">", "<", ">=", "<=", "==", "!=", "lambda"}
    
    def check_expr(expr: Expr, local_vars: Set[str]) -> bool:
        if isinstance(expr, Num):
            return True
        elif isinstance(expr, Var):
            return True
        elif isinstance(expr, Call):
            # '=' in expression
            if isinstance(expr.f, Var) and expr.f.name == "=":
                def collect_vars(e: Expr):
                    if isinstance(e, Var): local_vars.add(e.name)
                    elif isinstance(e, Call):
                        collect_vars(e.f)
                        for a in e.args: collect_vars(a)
                for arg in expr.args: collect_vars(arg)
                return True
            # 'lambda' in expression
            if isinstance(expr.f, Var) and expr.f.name == "lambda" and len(expr.args) >= 2:
                param = expr.args[0]
                if isinstance(param, Var):
                    new_locals = local_vars.copy()
                    new_locals.add(param.name)
                    return check_expr(expr.args[1], new_locals)
            
            return check_expr(expr.f, local_vars) and all(check_expr(arg, local_vars) for arg in expr.args)
        elif isinstance(expr, Block):
            current_locals = local_vars.copy()
            for e in expr.exprs:
                if not check_expr(e, current_locals): return False
                if isinstance(e, Call) and isinstance(e.f, Var) and e.f.name == "=":
                    if e.args and isinstance(e.args[0], Var):
                        current_locals.add(e.args[0].name)
            return True
        return True # Default for Atom/Apply

    # Top-level pass
    for expr in exprs:
        if isinstance(expr, Call) and isinstance(expr.f, Var) and expr.f.name == "=":
            if expr.args and isinstance(expr.args[0], Var):
                defined_vars.add(expr.args[0].name)
            
    for expr in exprs:
        if not check_expr(expr, set()): return False
    return True

def prune_token_tree(token_tree: undeterminable.Tree[Token]) -> undeterminable.Tree[Token]:
    paths = token_tree.flatten_paths()
    valid_paths = [p for p in paths if check_token_constraints(p)]
    if not valid_paths: return undeterminable.DeadEnd()
    return rebuild_tree(valid_paths)

def rebuild_tree(paths: List[List[Token]]) -> undeterminable.Tree[Token]:
    if not paths: return undeterminable.DeadEnd()
    branches = [build_path(p) for p in paths]
    return undeterminable.fork(branches)

def build_path(tokens: List[Token]) -> undeterminable.Tree[Token]:
    if not tokens: return undeterminable.Node([], [])
    return undeterminable.single(tokens[0], build_path(tokens[1:]))